src/Adapter/Carrier/Repository/CarrierRangeRepository.php

```diff
@@ -0,0 +1,234 @@
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Carrier\Repository;
+
+use Carrier;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Query\QueryBuilder;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\Exception\CarrierConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\Exception\CarrierException;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierId;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\CarrierRangesCollection;
+use PrestaShop\PrestaShop\Core\Domain\Carrier\ValueObject\ShippingMethod;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+
+/**
+ * Provides access to carrier range data source
+ */
+class CarrierRangeRepository
+{
+    public function __construct(
+        protected readonly Connection $connection,
+        protected readonly string $dbPrefix,
+        protected readonly CarrierRepository $carrierRepository,
+    ) {
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     *
+     * @return array<int, array<int|string>>
+     *
+     * @throws CarrierException
+     * @throws Exception
+     * @throws \PrestaShop\PrestaShop\Core\Domain\AttributeGroup\Attribute\Exception\AttributeNotFoundException
+     * @throws CarrierConstraintException
+     * @throws \PrestaShop\PrestaShop\Core\Domain\Zone\Exception\ZoneException
+     * @throws \PrestaShop\PrestaShop\Core\Exception\CoreException
+     */
+    public function get(CarrierId $carrierId, ShopConstraint $shopConstraint): array
+    {
+        // Check shop constraint
+        $this->assertShopConstraint($shopConstraint);
+
+        // Get carrier
+        $carrier = $this->carrierRepository->get($carrierId);
+
+        // Then, create query to retrieve carrier ranges
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('cd.id_zone, cr.delimiter1 AS range_from, cr.delimiter2 AS range_to, cd.price AS range_price')
+            ->from($this->dbPrefix . 'delivery', 'cd')
+            ->andWhere('cd.id_carrier = :carrierId')
+            ->setParameter('carrierId', $carrierId->getValue());
+
+        // Apply shipping method (weight or price) and shop constraint, then execute the query
+        $this->applyRangeTypeForQuery($qb, $carrier);
+        $this->applyShopConstraint($qb, $shopConstraint);
+
+        return $qb->executeQuery()->fetchAllAssociative();
+    }
+
+    /**
+     * @param CarrierId $carrierId
+     * @param CarrierRangesCollection $rangesCollection
+     * @param ShopConstraint $shopConstraint
+     */
+    public function set(CarrierId $carrierId, CarrierRangesCollection $rangesCollection, ShopConstraint $shopConstraint): void
```

Comment:
Strictly speaking we should keep the domain out of the repository, but we crossed that line long time ago simply by using `ProductId` `CategoryId` and all sort of VOs in the repository layer, so it's already too late anyway 😅